libraries/common/stream.effekt

```diff
@@ -0,0 +1,417 @@
+
+import array
+import bytearray
+import io/filesystem
+import io/error
+
+
+effect emit[A](value: A): Unit
+
+def for[A, R] { stream: () => R / emit[A] } { action: A => Unit }: R =
+  try {
+    stream()
+  } with emit[A] { value =>
+    resume(action(value))
+  }
+
+def for[A] { stream: () => Unit / emit[A] } { action: A => Unit }: Unit =
+  for[A, Unit]{stream}{action}
+
+def continually[A] { action: () => A }: Nothing / emit[A] = {
+  def go(): Nothing = {
+    do emit(action())
+    go()
+  }
+  go()
+}
+
+
+def each[A](list: List[A]): Unit / emit[A] =
+  list match {
+    case Nil() => ()
+    case Cons(head, tail) =>
+      do emit(head)
+      each(tail)
+  }
+
+def each[T](array: Array[T]): Unit / emit[T] = {
+  val n = array.size
+  def go(i: Int): Unit = {
+    if (i < n) {
+      do emit(array.unsafeGet(i))
+      go(i + 1)
+    }
+  }
+  go(0)
+}
+
+def each(bytes: ByteArray): Unit / emit[Byte] = {
+  val n = bytes.size
+  def go(i: Int): Unit = {
+    if (i < n) {
+      do emit(bytes.unsafeGet(i))
+      go(i + 1)
+    }
+  }
+  go(0)
+}
+
+// not option
+// not dequeue
+// not queue
+// not result
+// not seq
+
+
+effect stop(): Nothing
+
+
+def boundary[R] { program: () => R / stop }: Option[R] =
+  try {
+    Some(program())
+  } with stop {
+    def stop() = None()
+  }
+
+def exhaustively[A] { program: () => A / stop } { action: A => Unit }: Unit =
+  try {
+    def go(): Unit = {
+      action(program())
+      go()
+    }
+    go()
+  } with stop {
+    def stop() = ()
+  }
+
+record Indexed[A](index: Int, value: A)
+
+// In Effekt lower bounds are inclusive and upper bounds are exclusive
+
+def range(lower: Int, upper: Int): Unit / emit[Int] =
+  if (lower < upper) {
+    do emit(lower)
+    range(lower + 1, upper)
+  }
+
+def rangeTo(upper: Int): Unit / emit[Int] =
+  range(0, upper)
+
+def rangeFrom(lower: Int): Unit / emit[Int] = {
+  do emit(lower)
+  rangeFrom(lower + 1)
+}
+
+
+def index[A, R] { stream: () => R / emit[A] }: R / emit[Indexed[A]] = {
+  var i = 0;
+  try {
+    stream()
+  } with emit[A] { v =>
+    val c = i;
+    i = c + 1;
+    resume(do emit(Indexed(c, v)))
+  }
+}
+
+
+def take[A, R](number: Int) { stream: () => R / emit[A] }: R / { emit[A], stop } = {
+  var i = 0;
+  try {
+    stream()
+  } with emit[A] { v =>
+    val c = i;
+    if (c < number) {
+      i = c + 1;
+      resume(do emit(v))
+    } else {
+      do stop()
+    }
+  }
+}
+
+
+// If `number` is zero or negative it does nothing (like `stream::take`)
+def replicate[A](number: Int) { action: () => A }: Unit / emit[A] =
+  if (number > 0) {
+    do emit(action())
+    replicate(number - 1) {action}
+  }
+
+
+def sum[R] { stream : () => R / emit[Int] }: (R, Int) = {
+  var s = 0;
+  try {
+    (stream(), s)
+  } with emit[Int] { v =>
+    s = s + v;
+    resume(())
+  }
+}
+
+def sum { stream : () => Unit / emit[Int] }: Int =
+  sum[Unit]{stream}.second
+
+def collectList[A, R] { stream: () => R / emit[A] }: (R, List[A]) =
```

Comment:
You know I am not a big fan of returning two things here. Same below.

Reply:
I agree... if I understand the design of having an overt generic `R` and supporting streams that step to it (that then, you get streams stepping to `Unit` for-free) then I think it necessitates duplicating all these methods here that want to *do* something with the overt `R`. Since when `stream` is `() => Unit / emit[A]` (probably the most common use case?) you're always getting back this empty tuple `((), [...])` instead of just `[...]`

I don't really see the use case for having an overt generic `R` to begin with, to be honest. A Stream seems like something that should always step to the unit type. But I also haven't really worked with explicit Stream APIs all that much.

It might be worth considering having an alias `Stream` to `() => Unit` and a `Stream[R]` to `() => R`? And implementing separate functions for `Stream`, which mostly are direct wrappers around the corresponding functions for `Stream[R]`, but discard the empty tuple in these cases?